### ISSUE TITLE
Inconsistent behavior of markevery between semilogx and semilogy plots

### ISSUE THREAD

**Issue by Naereen** [matplotlib/matplotlib#8356]
### Bug report
- The `markevery` option for `plot`, `semilogx`, `semilogy`, and `loglog` plotting methods has a weird difference of behavior between `plot / semilogy` and `semilogx / loglog`, when using `markevery = (float, float)`.

**Code for reproduction**

```python
import numpy as np
import matplotlib.pyplot as plt

# Generate some data
X = np.linspace(-10, 10, 1000)
Y = np.sin(X)**2

# Specify markers
markers = ['o', 'v', '^', '<', '>']

def markevery(i):
    # I tried a *lot* of values here, none were working for the 4 plots
    return (i / 50., 0.1)

plt.figure()

# First plot in normal plot
plt.subplot(221)
for i in range(10):
    plt.plot(X, Y + i / 10., marker=markers[i % len(markers)], markevery=markevery(i), label=str(i))
plt.legend()
plt.title("plt.plot(): markers are here!")

# Then plot in semilogx
plt.subplot(222)
for i in range(10):
    plt.semilogx(X, Y + i / 10., marker=markers[i % len(markers)], markevery=markevery(i), label=str(i))
plt.legend()
plt.title("plt.semilogx(): markers are NOT here!")

# And then semilogy
plt.subplot(223)
for i in range(10):
    plt.semilogy(X, Y + i / 10., marker=markers[i % len(markers)], markevery=markevery(i), label=str(i))
plt.legend()
plt.title("plt.semilogy(): markers are here!")

# Finally plot in loglog
plt.subplot(224)
for i in range(10):
    plt.loglog(X, Y + i / 10., marker=markers[i % len(markers)], markevery=markevery(i), label=str(i))
plt.legend()
plt.title("plt.loglog(): markers are NOT here!")

plt.suptitle("Weird behavior of markevery between plt.plot, plt.semilogx, plt.semilogy, plt.loglog")
plt.show()
```

**Actual outcome**
![bug_markevery](https://cloud.githubusercontent.com/assets/11994719/24153472/d9d513de-0e4e-11e7-9655-347fe9d8ef8a.png)

**Expected outcome**

- One would expect the markers to be located in a **similar fashion** for each of the four plots method, `plot`, `semilogx`, `semilogy`, and `loglog`. The two plots on the left have markers correctly located (well spacen, with a different starting point for the different curves), and the two plots on the right do not show markers.
- Note that I tried a lot of different values for the `markevery` parameters (here `markevery = (i / 50., 0.1)` for the various curves, with changing `i` to change the starting point of the markers),  and always encounter this inconsistency between left and right.

**Matplotlib version**

- Matplotlib version 2.0.0 (`2.0.0+3337.gc15694b`), with Python 3.5, on Ubuntu 16.10
- Python installed with apt, latest version of Matplotlib installed with `pip --upgrade`.

----

Maybe I don't understand the [semantic behind using `markevery = (float, float)`](http://matplotlib.org/api/lines_api.html#matplotlib.lines.Line2D.set_markevery), but as far as I can tell, this behavior is weird.
Thanks in advance!

**Comment by Naereen** [comment]
> By the way, the reason I am asking is because I use a similar trick of `markevery = (i / 50., 0.1)` in some research code showing half a dozen of plots in the same figure, and I am puzzled that it exhibits a different behavior between `plot` and `semilogy`: markers are perfectly displayed but not for `semilogy` plots...

**Comment by phobson** [comment]
Thanks for the detailed report. What happens if you use `plot` and set the x/yscales afterwards?

**Comment by Naereen** [comment]
Thanks for your quicky reply.
Well, it fixes the issue.

The corrected script can be:
```python
import numpy as np
import matplotlib.pyplot as plt

# Generate some data
X = np.linspace(-10, 10, 1000)
Y = np.sin(X)**2

# Specify markers
markers = ['o', 'v', '^', '<', '>']


def markevery(i):
    return (i / 50., 0.1)


fig = plt.figure()

# First plot in normal plot
ax = fig.add_subplot(2, 2, 1)
for i in range(10):
    plt.plot(X, Y + i / 10., marker=markers[i % len(markers)], markevery=markevery(i), label=str(i))
plt.legend()
plt.title("plt.plot(): markers are here!")

# Then plot in semilogx
ax = fig.add_subplot(2, 2, 2)
for i in range(10):
    plt.plot(X, Y + i / 10., marker=markers[i % len(markers)], markevery=markevery(i), label=str(i))
    ax.set_xscale('log')
plt.legend()
plt.title("plt.plot() with ax.set_xscale('log'): markers are NOW here! Fixed!")

# And then semilogy
ax = fig.add_subplot(2, 2, 3)
for i in range(10):
    plt.plot(X, Y + i / 10., marker=markers[i % len(markers)], markevery=markevery(i), label=str(i))
    ax.set_yscale('log')
plt.legend()
plt.title("plt.plot() with ax.set_yscale('log'): markers are here!")

# Finally plot in loglog
ax = fig.add_subplot(2, 2, 4)
for i in range(10):
    plt.plot(X, Y + i / 10., marker=markers[i % len(markers)], markevery=markevery(i), label=str(i))
    ax.set_xscale('log')
    ax.set_yscale('log')
plt.legend()
plt.title("plt.plot() with ax.set_xscale('log') and ax.set_yscale('log'): markers are NOW here! Fixed!")

plt.suptitle("Weird behavior of markevery between plt.plot, plt.semilogx, plt.semilogy, plt.loglog\nhttps://github.com/matplotlib/matplotlib/issues/8356")
plt.show()
```

And it gives what the first one should have given:
![bug_markevery_fixed](https://cloud.githubusercontent.com/assets/11994719/24158958/1a3049b2-0e5e-11e7-8e99-8687646c2135.png)

**Comment by phobson** [comment]
OK. Thanks, that'll help us track it down.

**Comment by Naereen** [comment]
Is there a way to use `set_xscale` or `set_yscale` without having to store and use the figure and then the axis object?

**Comment by phobson** [comment]
Yes, but I always recommend that you explicitly create and store your Figure and Axes.

You can do e.g., `plt.xscale('log')`, but then you have keep track in your head which Axes you're operating on. Kind of a mental burden, IMHO.

**Comment by phobson** [comment]
For instance, I would change your code to be

```
fig, axes = pyplot.subplots(nrows=2, ncols=2)
axes[0, 0].plot(...)
axes[0, 0].set_xscale('log')
# yada yada
```

alternatively, you could do:
```
fig, ((ax1, ax2), (ax3, ax4))= pyplot.subplots(nrows=2, ncols=2)
ax1.plot(...)
ax1.set_xscale('log')
# yada yada
```

**Comment by Naereen** [comment]
> OK, I can confirm that setting the `xscale`/`yscale` manually instead of using `plt.semilogx`/`plt.semilogy` works also for my research code.
> ![fixed](https://cloud.githubusercontent.com/assets/11994719/24161429/ad06a744-0e64-11e7-858c-bf2ca4987556.png)

**Comment by Naereen** [comment]
If you feel like you need help, I will try to fix the issue myself on matplotlib's code, if someone points me to the exact and correct function/part that needs to be modified. Don't hesitate to ask me, I will try my best!

**Comment by phobson** [comment]
Start digging at e.g., https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/axes/_axes.py#L1551

**Comment by Naereen** [comment]
Seems to me that, for safety, the `self.set_xscale('log', **d)` could also be done after the `l = self.plot(*args, **kwargs)` ?
Doing this (hacky) change, for semilogx,semilogy & loglog, will fix the issues.

**Comment by Naereen** [comment]
In fact, I am not proper way to use `set_x/yscale` is *before* the call to plot.
Could someone confirm?

**Comment by phobson** [comment]
@Naereen that all sounds right to me.

**Comment by Naereen** [comment]
I (dirtily) tried to apply my suggestion on my system-wide matplotlib, and it doesn't fix anything.
For instance for `semilogx` in `_axes.py`:
```python
        # Lines 1650 something
        self.set_yscale('log', **d)
        b = self._hold
        self._hold = True  # we've already processed the hold

        l = self.plot(*args, **kwargs)
        print("Warning: using my hack of setting yscale to log after self.plot(...)...")
        self.set_yscale('log', **d)  # Experimental?
        self._hold = b  # restore the hold
        return l
```
Still produces the same figure.

But changing in `pyplot.py`, `semilogx`, `semilogy` and `loglog`, by adding, for instance (line 3108),
```python
    try:
        ret = ax.loglog(*args, **kwargs)
        ax.set_xscale('log')  # Experimental!
        ax.set_yscale('log')  # Experimental!
    finally:
        ...
```

**Comment by Naereen** [comment]
So I guess my suggestion is not good, and there is definitely a weird bug! :worried:

The second experimental modification (in `pyploy.py`) should *not* be done, as `ax.set_xscale('log')` in this case does not use the `xbase` (and others) optional keywords arguments...

**Comment by tacaswell** [comment]
I think the underlying problem is that when using the float version of markevery, the data space <-> screen space transform is used to back out which points should be labeled.  If you change the scale the previous computation end up being very wrong.

This belongs to a class of log-related bugs for autoscaling as well.

We probably need a notify mechanism when the scale changes to let artists know they should re-compute internal state.

Also suspect there may be 'race conditions' here where a `draw` forces some things to update so some of the order related changes may be related to that.

**Comment by Naereen** [comment]
OK. Clearly, I don't enough about internals of matplotlib to really understand what you explained, but I'm glad to see you seem to know what's going on.
Thanks, and good luck!

**Comment by Naereen** [comment]
Will this be included for the 2.1 release?

**Comment by phobson** [comment]
@Naereen we've milestoned it to be fixed by 2.1, but that's not a 100% guarantee.

**Comment by Naereen** [comment]
Hi,
Any update? Any help needed?

**Comment by tacaswell** [comment]
Any help working on this would be appreciated :)  

Does setting `markevery` on a line object correctly recompute the locations with the current scale?

**Comment by Naereen** [comment]
Yes

**Comment by ksunden** [comment]
This seems to have been fixed somewhere along the way, not too sure where:

![Figure_1](https://user-images.githubusercontent.com/2501846/230823060-30e811e0-96f5-4003-84ba-52fa4b0b7e14.png)

**Comment by QuLogic** [comment]
This was fixed by #9477 (because the `markevery` calculation produces all `nan` and thus plots nowhere.)
